Tolerance too strict in _test_node_vertex_consistency_resampling

We have seen lots of errors like these:

```
"/Users/mdiener/Work/efuse/meshmode/meshmode/mesh/__init__.py", line 1207, in _test_node_vertex_consistency
    assert _test_node_vertex_consistency_resampling(mesh, mgrp, tol)
  File "/Users/mdiener/Work/efuse/meshmode/meshmode/mesh/__init__.py", line 1195, in _test_node_vertex_consistency_resampling
    assert max_el_vertex_error < tol*size, max_el_vertex_error
AssertionError: 1.144170793882154e-15
```

This is the code:
https://github.com/inducer/meshmode/blob/48abe4bedf5e54e57fca392756423b6f45abff9a/meshmode/mesh/__init__.py#L1178

--- a/meshmode/__init__.py
+++ b/meshmode/__init__.py
@@ -25,6 +25,7 @@ __doc__ = """
 .. exception:: Error
 .. exception:: DataUnavailable
 .. exception:: FileExistsError
+.. exception:: InconsistentVerticesError
 """
 
 from builtins import FileExistsError  # noqa: F401
@@ -36,6 +37,14 @@ class Error(RuntimeError):
 
 
 class DataUnavailable(Error):
+    pass
+
+
+class InconsistentVerticesError(Error):
+    """
+    Raised when an element's local-to-global mapping does not map the unit vertices
+    to the corresponding values in the mesh's *vertices* array.
+    """
     pass
 
 

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1233,13 +1233,17 @@ def _test_node_vertex_consistency_resampling(mesh, mgrp, tol):
     if tol is None:
         tol = 1e3 * np.finfo(per_element_vertex_errors.dtype).eps
 
-    from meshmode.mesh.processing import find_bounding_box
+    grp_vertices = mesh.vertices[:, mgrp.vertex_indices]
 
-    bbox_min, bbox_max = find_bounding_box(mesh)
-    size = la.norm(bbox_max-bbox_min)
+    per_element_sizes = la.norm(
+        np.max(grp_vertices, axis=-1)
+        - np.min(grp_vertices, axis=-1),
+        axis=0)
 
-    max_el_vertex_error = np.max(per_element_vertex_errors)
-    assert max_el_vertex_error < tol*size, max_el_vertex_error
+    per_element_tols = tol * np.maximum(per_element_sizes, 1)
+
+    assert np.all(per_element_vertex_errors < per_element_tols), \
+        per_element_vertex_errors[per_element_vertex_errors >= per_element_tols][0]
 
     return True
 

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1218,9 +1218,11 @@ def _mesh_group_node_vertex_error(mesh, mgrp):
     return map_vertices - grp_vertices
 
 
-def _test_node_vertex_consistency_resampling(mesh, mgrp, tol):
+def _test_node_vertex_consistency_resampling(mesh, igrp, tol):
     if mesh.vertices is None:
         return True
+
+    mgrp = mesh.groups[igrp]
 
     if mgrp.nelements == 0:
         return True
@@ -1243,8 +1245,15 @@ def _test_node_vertex_consistency_resampling(mesh, mgrp, tol):
 
     per_element_tols = abs_tol + rel_tol * coord_scales
 
-    assert np.all(per_element_vertex_errors < per_element_tols), \
-        per_element_vertex_errors[per_element_vertex_errors >= per_element_tols][0]
+    elements_above_tol, = np.where(per_element_vertex_errors >= per_element_tols)
+    if len(elements_above_tol) > 0:
+        i_grp_elem = elements_above_tol[0]
+        ielem = i_grp_elem + mesh.base_element_nrs[igrp]
+        from meshmode import InconsistentVerticesError
+        raise AssertionError(
+            f"vertex consistency check failed for element {ielem}; "
+            f"{per_element_vertex_errors[i_grp_elem]} >= "
+            f"{per_element_tols[i_grp_elem]}")
 
     return True
 
@@ -1254,9 +1263,9 @@ def _test_node_vertex_consistency(mesh, tol):
     unit vertices.
     """
 
-    for mgrp in mesh.groups:
+    for igrp, mgrp in enumerate(mesh.groups):
         if isinstance(mgrp, _ModepyElementGroup):
-            assert _test_node_vertex_consistency_resampling(mesh, mgrp, tol)
+            assert _test_node_vertex_consistency_resampling(mesh, igrp, tol)
         else:
             from warnings import warn
             warn("not implemented: node-vertex consistency check for '%s'"

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1250,7 +1250,7 @@ def _test_node_vertex_consistency_resampling(mesh, igrp, tol):
         i_grp_elem = elements_above_tol[0]
         ielem = i_grp_elem + mesh.base_element_nrs[igrp]
         from meshmode import InconsistentVerticesError
-        raise AssertionError(
+        raise InconsistentVerticesError(
             f"vertex consistency check failed for element {ielem}; "
             f"{per_element_vertex_errors[i_grp_elem]} >= "
             f"{per_element_tols[i_grp_elem]}")

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1224,26 +1224,23 @@ def _test_node_vertex_consistency_resampling(mesh, igrp, tol):
 
     mgrp = mesh.groups[igrp]
 
-    if mgrp.nelements == 0:
+    if mgrp.dim == 0:
+        return True
+    elif mgrp.nelements == 0:
         return True
 
     per_vertex_errors = _mesh_group_node_vertex_error(mesh, mgrp)
-    per_element_vertex_errors = np.sqrt(
-        np.sum(np.sum((per_vertex_errors)**2, axis=0), axis=-1)
-    )
+    per_element_vertex_errors = np.max(
+        np.max(np.abs(per_vertex_errors), axis=-1), axis=0)
 
     if tol is None:
         tol = 1e3 * np.finfo(per_element_vertex_errors.dtype).eps
-
-    # Assume these are the same for now; can be made separate arguments if necessary
-    abs_tol = tol
-    rel_tol = tol
 
     grp_vertices = mesh.vertices[:, mgrp.vertex_indices]
 
     coord_scales = np.max(np.max(np.abs(grp_vertices), axis=-1), axis=0)
 
-    per_element_tols = abs_tol + rel_tol * coord_scales
+    per_element_tols = tol * coord_scales
 
     elements_above_tol, = np.where(per_element_vertex_errors >= per_element_tols)
     if len(elements_above_tol) > 0:

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1224,9 +1224,7 @@ def _test_node_vertex_consistency_resampling(mesh, igrp, tol):
 
     mgrp = mesh.groups[igrp]
 
-    if mgrp.dim == 0:
-        return True
-    elif mgrp.nelements == 0:
+    if mgrp.nelements == 0:
         return True
 
     per_vertex_errors = _mesh_group_node_vertex_error(mesh, mgrp)
@@ -1240,7 +1238,7 @@ def _test_node_vertex_consistency_resampling(mesh, igrp, tol):
 
     coord_scales = np.max(np.max(np.abs(grp_vertices), axis=-1), axis=0)
 
-    per_element_tols = tol * coord_scales
+    per_element_tols = tol + tol * coord_scales
 
     elements_above_tol, = np.where(per_element_vertex_errors >= per_element_tols)
     if len(elements_above_tol) > 0:

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1259,6 +1259,8 @@ def _test_node_vertex_consistency(mesh, tol):
     """Ensure that order of by-index vertices matches that of mapped
     unit vertices.
     """
+    if not __debug__:
+        return True
 
     for igrp, mgrp in enumerate(mesh.groups):
         if isinstance(mgrp, _ModepyElementGroup):

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1240,6 +1240,8 @@ def _test_node_vertex_consistency_resampling(mesh, mgrp, tol):
         - np.min(grp_vertices, axis=-1),
         axis=0)
 
+    # Only scale by the element size if the elements are sufficiently large,
+    # to prevent the tolerance from dropping below rounding error
     per_element_tols = tol * np.maximum(per_element_sizes, 1)
 
     assert np.all(per_element_vertex_errors < per_element_tols), \

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1230,13 +1230,12 @@ def _test_node_vertex_consistency_resampling(mesh, mgrp, tol):
         np.sum(np.sum((per_vertex_errors)**2, axis=0), axis=-1)
     )
 
-    abs_tol = 1e-12
-    rel_tol = 1e-7
-    if tol is not None:
-        # Rescale for compatibility with tolerances from old version that used
-        # mesh bounding box for scale
-        old_default_tol = 1e3*np.finfo(per_element_vertex_errors.dtype).eps
-        rel_tol *= tol/old_default_tol
+    if tol is None:
+        tol = 1e3 * np.finfo(per_element_vertex_errors.dtype).eps
+
+    # Assume these are the same for now; can be made separate arguments if necessary
+    abs_tol = tol
+    rel_tol = tol
 
     grp_vertices = mesh.vertices[:, mgrp.vertex_indices]
 

--- a/meshmode/mesh/__init__.py
+++ b/meshmode/mesh/__init__.py
@@ -1230,19 +1230,19 @@ def _test_node_vertex_consistency_resampling(mesh, mgrp, tol):
         np.sum(np.sum((per_vertex_errors)**2, axis=0), axis=-1)
     )
 
-    if tol is None:
-        tol = 1e3 * np.finfo(per_element_vertex_errors.dtype).eps
+    abs_tol = 1e-12
+    rel_tol = 1e-7
+    if tol is not None:
+        # Rescale for compatibility with tolerances from old version that used
+        # mesh bounding box for scale
+        old_default_tol = 1e3*np.finfo(per_element_vertex_errors.dtype).eps
+        rel_tol *= tol/old_default_tol
 
     grp_vertices = mesh.vertices[:, mgrp.vertex_indices]
 
-    per_element_sizes = la.norm(
-        np.max(grp_vertices, axis=-1)
-        - np.min(grp_vertices, axis=-1),
-        axis=0)
+    coord_scales = np.max(np.max(np.abs(grp_vertices), axis=-1), axis=0)
 
-    # Only scale by the element size if the elements are sufficiently large,
-    # to prevent the tolerance from dropping below rounding error
-    per_element_tols = tol * np.maximum(per_element_sizes, 1)
+    per_element_tols = abs_tol + rel_tol * coord_scales
 
     assert np.all(per_element_vertex_errors < per_element_tols), \
         per_element_vertex_errors[per_element_vertex_errors >= per_element_tols][0]

--- a/test/test_mesh.py
+++ b/test/test_mesh.py
@@ -1139,21 +1139,99 @@ def test_node_vertex_consistency_check(actx_factory):
     from meshmode import InconsistentVerticesError
 
     # Mesh bounds invariance
+
+    def gen_rect_mesh_with_perturbed_vertices(
+            a, b, nelements_per_axis, perturb_amount):
+        mesh_unperturbed = mgen.generate_regular_rect_mesh(
+            a=a, b=b, nelements_per_axis=nelements_per_axis)
+        return mesh_unperturbed.copy(  # noqa: F841
+            vertices=(
+                mesh_unperturbed.vertices
+                + perturb_amount*np.ones(mesh_unperturbed.vertices.shape)))
+
+    # Find critical perturbation amount "w" such that vertices shifted by w works
+    # but 10*w doesn't
     h = 1e-5
-    for p in range(5):
-        nelems = 2**p
+    nelems = 8
+    size = h*nelems
+    crit_perturb = None
+    for p in range(32):
+        w = h*10**(p-16)
+        try:
+            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+                a=(-size/2,), b=(size/2,), nelements_per_axis=(nelems,),
+                perturb_amount=w)
+        except InconsistentVerticesError:
+            if p > 0:
+                crit_perturb = w/10
+            break
+    if crit_perturb is None:
+        raise RuntimeError("failed to find critical vertex perturbation amount")
+
+    # Scale invariance
+    nelems = 8
+    for p in range(32):
+        h = 10**(p-16)
         size = h*nelems
-        mesh = mgen.generate_regular_rect_mesh(  # noqa: F841
-            a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3)
+        perturb = crit_perturb*h/1e-5
+        mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+            a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
+            perturb_amount=perturb)
+        with pytest.raises(InconsistentVerticesError):
+            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+                a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
+                perturb_amount=10*perturb)
 
     # Translation invariance
     h = 1e-5
     nelems = 8
     size = h*nelems
+    for p in range(10):
+        shift = 10**p-1
+        perturb = crit_perturb*(size/2 + shift)/(size/2)
+        mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+            a=(shift-size/2,)*3, b=(shift+size/2,)*3,
+            nelements_per_axis=(nelems,)*3,
+            perturb_amount=perturb)
+        with pytest.raises(InconsistentVerticesError):
+            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+                a=(shift-size/2,)*3, b=(shift+size/2,)*3,
+                nelements_per_axis=(nelems,)*3,
+                perturb_amount=10*perturb)
+
+    # Aspect ratio invariance
+    h = 1e-5
+    nelems = 8
+    size = h*nelems
+    for p in range(10):
+        h_x = 10**p * h
+        size_x = h_x * nelems
+        perturb = crit_perturb*h_x/1e-5
+        mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+            a=(-size_x/2,) + (-size/2,)*2,
+            b=(size_x/2,) + (size/2,)*2,
+            nelements_per_axis=(nelems,)*3,
+            perturb_amount=perturb)
+        with pytest.raises(InconsistentVerticesError):
+            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+                a=(-size_x/2,) + (-size/2,)*2,
+                b=(size_x/2,) + (size/2,)*2,
+                nelements_per_axis=(nelems,)*3,
+                perturb_amount=10*perturb)
+
+    # Mesh size relative to element size invariance
+    h = 1e-5
     for p in range(5):
-        mesh = mgen.generate_regular_rect_mesh(  # noqa: F841
-            a=(10**p-size/2,)*3, b=(10**p+size/2,)*3,
-            nelements_per_axis=(nelems,)*3)
+        nelems = 2**(5-p)
+        size = h*nelems
+        perturb = crit_perturb
+        mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+            a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
+            perturb_amount=perturb)
+        with pytest.raises(InconsistentVerticesError):
+            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+                a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
+                perturb_amount=10*perturb)
 
     # Zero-D elements
     h = 1e-5
@@ -1183,7 +1261,7 @@ def test_node_vertex_consistency_check(actx_factory):
     faces_mesh = make_face_restriction(  # noqa: F841
         actx, vol_discr, group_factory, FACE_RESTR_ALL, per_face_groups=False)
 
-    # Expected failing case: element vertex indices rotated
+    # Element vertex indices rotated
     with pytest.raises(InconsistentVerticesError):
         vol_mesh_unrotated = mgen.generate_regular_rect_mesh(
             a=(-1,)*2, b=(1,)*2,

--- a/test/test_mesh.py
+++ b/test/test_mesh.py
@@ -24,6 +24,7 @@ THE SOFTWARE.
 """
 
 from functools import partial
+from dataclasses import replace
 import numpy as np
 import numpy.linalg as la
 import pytest
@@ -1179,6 +1180,18 @@ def test_node_vertex_consistency_check(actx_factory):
     vol_discr = Discretization(actx, vol_mesh, group_factory)
     faces_mesh = make_face_restriction(  # noqa: F841
         actx, vol_discr, group_factory, FACE_RESTR_ALL, per_face_groups=False)
+
+    # Expected failing case: element vertex indices rotated
+    with pytest.raises(AssertionError):
+        vol_mesh_unrotated = mgen.generate_regular_rect_mesh(
+            a=(-1,)*2, b=(1,)*2,
+            nelements_per_axis=(8,)*2)
+        vol_mesh = vol_mesh_unrotated.copy(  # noqa: F841
+            groups=[
+                replace(
+                    grp,
+                    vertex_indices=np.roll(grp.vertex_indices, 1, axis=1))
+                for grp in vol_mesh_unrotated.groups])
 
 # }}}
 

--- a/test/test_mesh.py
+++ b/test/test_mesh.py
@@ -1130,6 +1130,59 @@ def test_mesh_element_group_constructor(shape_cls, group_cls):
 # }}}
 
 
+# {{{ test_node_vertex_consistency_check
+
+def test_node_vertex_consistency_check(actx_factory):
+    actx = actx_factory()
+
+    # Mesh bounds invariance
+    h = 1e-5
+    for p in range(5):
+        nelems = 2**p
+        size = h*nelems
+        mesh = mgen.generate_regular_rect_mesh(  # noqa: F841
+            a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3)
+
+    # Translation invariance
+    h = 1e-5
+    nelems = 8
+    size = h*nelems
+    for p in range(5):
+        mesh = mgen.generate_regular_rect_mesh(  # noqa: F841
+            a=(10**p-size/2,)*3, b=(10**p+size/2,)*3,
+            nelements_per_axis=(nelems,)*3)
+
+    # Zero-size elements
+    h = 1e-5
+    nelems = 7
+    size = h*nelems
+    vol_mesh = mgen.generate_regular_rect_mesh(
+        a=(-size/2,), b=(size/2,),
+        nelements_per_axis=(nelems,))
+    from meshmode.discretization import Discretization
+    group_factory = default_simplex_group_factory(1, 1)
+    vol_discr = Discretization(actx, vol_mesh, group_factory)
+    from meshmode.discretization.connection import (
+        FACE_RESTR_ALL,
+        make_face_restriction)
+    faces_mesh = make_face_restriction(  # noqa: F841
+        actx, vol_discr, group_factory, FACE_RESTR_ALL, per_face_groups=False)
+
+    # Zero-size elements at the origin
+    h = 1e-5
+    nelems = 8
+    size = h*nelems
+    vol_mesh = mgen.generate_regular_rect_mesh(
+        a=(-size/2,), b=(size/2,),
+        nelements_per_axis=(nelems,))
+    group_factory = default_simplex_group_factory(1, 1)
+    vol_discr = Discretization(actx, vol_mesh, group_factory)
+    faces_mesh = make_face_restriction(  # noqa: F841
+        actx, vol_discr, group_factory, FACE_RESTR_ALL, per_face_groups=False)
+
+# }}}
+
+
 # {{{ mesh boundary gluing
 
 @pytest.mark.parametrize("use_tree", [False, True])

--- a/test/test_mesh.py
+++ b/test/test_mesh.py
@@ -1138,6 +1138,9 @@ def test_node_vertex_consistency_check(actx_factory):
 
     from meshmode import InconsistentVerticesError
 
+    dtype = np.float64
+    tol = 1e3 * np.finfo(dtype).eps
+
     # Mesh bounds invariance
 
     def gen_rect_mesh_with_perturbed_vertices(
@@ -1147,18 +1150,19 @@ def test_node_vertex_consistency_check(actx_factory):
         return mesh_unperturbed.copy(  # noqa: F841
             vertices=(
                 mesh_unperturbed.vertices
-                + perturb_amount*np.ones(mesh_unperturbed.vertices.shape)))
+                + perturb_amount*np.ones(mesh_unperturbed.vertices.shape)),
+            node_vertex_consistency_tolerance=tol)
 
     # Find critical perturbation amount "w" such that vertices shifted by w works
     # but 10*w doesn't
-    h = 1e-5
+    h = 1
     nelems = 8
     size = h*nelems
     crit_perturb = None
     for p in range(32):
         w = h*10**(p-16)
         try:
-            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+            gen_rect_mesh_with_perturbed_vertices(
                 a=(-size/2,), b=(size/2,), nelements_per_axis=(nelems,),
                 perturb_amount=w)
         except InconsistentVerticesError:
@@ -1173,68 +1177,69 @@ def test_node_vertex_consistency_check(actx_factory):
     for p in range(32):
         h = 10**(p-16)
         size = h*nelems
-        perturb = crit_perturb*h/1e-5
-        mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+        perturb = crit_perturb*h
+        gen_rect_mesh_with_perturbed_vertices(
             a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
             perturb_amount=perturb)
-        with pytest.raises(InconsistentVerticesError):
-            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
-                a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
-                perturb_amount=10*perturb)
+        if 10*perturb > tol:
+            with pytest.raises(InconsistentVerticesError):
+                gen_rect_mesh_with_perturbed_vertices(
+                    a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
+                    perturb_amount=10*perturb)
 
     # Translation invariance
-    h = 1e-5
+    h = 1
     nelems = 8
     size = h*nelems
     for p in range(10):
         shift = 10**p-1
         perturb = crit_perturb*(size/2 + shift)/(size/2)
-        mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+        gen_rect_mesh_with_perturbed_vertices(
             a=(shift-size/2,)*3, b=(shift+size/2,)*3,
             nelements_per_axis=(nelems,)*3,
             perturb_amount=perturb)
         with pytest.raises(InconsistentVerticesError):
-            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+            gen_rect_mesh_with_perturbed_vertices(
                 a=(shift-size/2,)*3, b=(shift+size/2,)*3,
                 nelements_per_axis=(nelems,)*3,
                 perturb_amount=10*perturb)
 
     # Aspect ratio invariance
-    h = 1e-5
+    h = 1
     nelems = 8
     size = h*nelems
     for p in range(10):
         h_x = 10**p * h
         size_x = h_x * nelems
-        perturb = crit_perturb*h_x/1e-5
-        mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+        perturb = crit_perturb*h_x
+        gen_rect_mesh_with_perturbed_vertices(
             a=(-size_x/2,) + (-size/2,)*2,
             b=(size_x/2,) + (size/2,)*2,
             nelements_per_axis=(nelems,)*3,
             perturb_amount=perturb)
         with pytest.raises(InconsistentVerticesError):
-            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+            gen_rect_mesh_with_perturbed_vertices(
                 a=(-size_x/2,) + (-size/2,)*2,
                 b=(size_x/2,) + (size/2,)*2,
                 nelements_per_axis=(nelems,)*3,
                 perturb_amount=10*perturb)
 
     # Mesh size relative to element size invariance
-    h = 1e-5
+    h = 1
     for p in range(5):
         nelems = 2**(5-p)
         size = h*nelems
         perturb = crit_perturb
-        mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+        gen_rect_mesh_with_perturbed_vertices(
             a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
             perturb_amount=perturb)
         with pytest.raises(InconsistentVerticesError):
-            mesh = gen_rect_mesh_with_perturbed_vertices(  # noqa: F841
+            gen_rect_mesh_with_perturbed_vertices(
                 a=(-size/2,)*3, b=(size/2,)*3, nelements_per_axis=(nelems,)*3,
                 perturb_amount=10*perturb)
 
     # Zero-D elements
-    h = 1e-5
+    h = 1
     nelems = 7
     size = h*nelems
     vol_mesh = mgen.generate_regular_rect_mesh(
@@ -1246,11 +1251,11 @@ def test_node_vertex_consistency_check(actx_factory):
     from meshmode.discretization.connection import (
         FACE_RESTR_ALL,
         make_face_restriction)
-    faces_mesh = make_face_restriction(  # noqa: F841
+    make_face_restriction(
         actx, vol_discr, group_factory, FACE_RESTR_ALL, per_face_groups=False)
 
     # Zero-D elements at the origin
-    h = 1e-5
+    h = 1
     nelems = 8
     size = h*nelems
     vol_mesh = mgen.generate_regular_rect_mesh(
@@ -1258,7 +1263,7 @@ def test_node_vertex_consistency_check(actx_factory):
         nelements_per_axis=(nelems,))
     group_factory = default_simplex_group_factory(1, 1)
     vol_discr = Discretization(actx, vol_mesh, group_factory)
-    faces_mesh = make_face_restriction(  # noqa: F841
+    make_face_restriction(
         actx, vol_discr, group_factory, FACE_RESTR_ALL, per_face_groups=False)
 
     # Element vertex indices rotated

--- a/test/test_mesh.py
+++ b/test/test_mesh.py
@@ -1136,6 +1136,8 @@ def test_mesh_element_group_constructor(shape_cls, group_cls):
 def test_node_vertex_consistency_check(actx_factory):
     actx = actx_factory()
 
+    from meshmode import InconsistentVerticesError
+
     # Mesh bounds invariance
     h = 1e-5
     for p in range(5):
@@ -1182,7 +1184,7 @@ def test_node_vertex_consistency_check(actx_factory):
         actx, vol_discr, group_factory, FACE_RESTR_ALL, per_face_groups=False)
 
     # Expected failing case: element vertex indices rotated
-    with pytest.raises(AssertionError):
+    with pytest.raises(InconsistentVerticesError):
         vol_mesh_unrotated = mgen.generate_regular_rect_mesh(
             a=(-1,)*2, b=(1,)*2,
             nelements_per_axis=(8,)*2)

--- a/test/test_mesh.py
+++ b/test/test_mesh.py
@@ -1153,7 +1153,7 @@ def test_node_vertex_consistency_check(actx_factory):
             a=(10**p-size/2,)*3, b=(10**p+size/2,)*3,
             nelements_per_axis=(nelems,)*3)
 
-    # Zero-size elements
+    # Zero-D elements
     h = 1e-5
     nelems = 7
     size = h*nelems
@@ -1169,7 +1169,7 @@ def test_node_vertex_consistency_check(actx_factory):
     faces_mesh = make_face_restriction(  # noqa: F841
         actx, vol_discr, group_factory, FACE_RESTR_ALL, per_face_groups=False)
 
-    # Zero-size elements at the origin
+    # Zero-D elements at the origin
     h = 1e-5
     nelems = 8
     size = h*nelems

